crates/runtime/src/datafusion/param_utils.rs

```diff
@@ -0,0 +1,269 @@
+///
+/// # Supported JSON Formats
+/// - Array: Converted to positional parameters (ParamValues::Positional).
+/// - Object: Converted to named parameters (ParamValues::Named).
+/// - Primitive types (null, bool, number, string) are supported within arrays or objects.
+///
+/// # Errors
+/// - Returns an error if the top-level JSON is not an array or object.
+/// - Returns an error if any value cannot be converted to a ScalarValue.
+pub fn convert_json_to_param_values(json: Value) -> Result<ParamValues, String> {
+    match json {
+        Value::Array(arr) => {
+            let mut vec = Vec::with_capacity(arr.len());
+            for (i, val) in arr.into_iter().enumerate() {
+                let scalar = json_to_scalar(&val).map_err(|e| {
+                    format!("failed to convert array element at index {}: {}", i, e)
+                })?;
+                vec.push(scalar);
+            }
+            Ok(ParamValues::List(vec))
+        }
+        Value::Object(obj) => {
+            let mut map = HashMap::new();
+            for (key, val) in obj {
+                let scalar = json_to_scalar(&val)
+                    .map_err(|e| format!("failed to convert value for key '{}': {}", key, e))?;
+                map.insert(key, scalar);
+            }
+            Ok(ParamValues::Map(map))
+        }
+        _ => Err("params must be a JSON array or object".to_string()),
+    }
+}
+
+/// Helper function to convert a single serde_json::Value to a DataFusion ScalarValue.
+///
+/// # Arguments
+/// * `json` - The JSON value to convert.
+///
+/// # Returns
+/// * `Result<ScalarValue, String>` - The converted ScalarValue or an error message.
+///
+/// # Supported JSON Types
+/// - Null -> ScalarValue::Utf8(None)
+/// - Bool -> ScalarValue::Boolean(Some(bool))
+/// - Number -> ScalarValue::Int64(Some(i64)) if integer, else ScalarValue::Float64(Some(f64))
+/// - String -> ScalarValue::Utf8(Some(String))
+/// - Arrays and objects return an error (handled at a higher level).
+fn json_to_scalar(json: &Value) -> Result<ScalarValue, String> {
+    match json {
+        Value::Null => Ok(ScalarValue::Utf8(None)),
+        Value::Bool(b) => Ok(ScalarValue::Boolean(Some(*b))),
+        Value::Number(n) => {
+            if let Some(i) = n.as_i64() {
+                Ok(ScalarValue::Int64(Some(i)))
+            } else if let Some(f) = n.as_f64() {
+                Ok(ScalarValue::Float64(Some(f)))
+            } else {
+                Err("Unsupported JSON number format".to_string())
+            }
+        }
+        Value::String(s) => Ok(ScalarValue::Utf8(Some(s.clone()))),
+        Value::Array(_) | Value::Object(_) => {
+            Err("Nested arrays or objects are not supported as parameter values".to_string())
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use datafusion::scalar::ScalarValue;
+    use serde_json::json;
+    use std::collections::HashMap;
+
+    fn scalar_to_string(scalar: &ScalarValue) -> String {
```

Comment:
I think it could be just  `scalar.to_string()`, no need for manual conversion to string